src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template

```diff
@@ -91,12 +89,12 @@ class Direct$Type$Buffer$RW$$BO$
         }
 
         public void run() {
-            if (address == 0) {
+            if (address == 0 || deallocated) {
                 // Paranoia
                 return;
             }
+            deallocated = true;
```

Comment:
> Shouldn't this be an atomic compareAndSet instead? Thread A can read deallocated and pause before the field write, then thread B read deallocated, continue to set the field. There will be a double free when thread A resumes.

Deallocator is the cleaner action so it should only ever be run once by the cleaner thread. There is a back door via Unsafe.invokeCleaner for freeing a buffer potentially in use but the cleaner action should still only be invoked once.  The "Paranoia" comment and guarding against being called again may date back to the older cleaner code, requires a bit of digging to see if (and the flag) can be removed.

Reply:
If only one cleaner thread runs, can we just safely assume it is run only once and simply remove the `deallocated` check, which only defends against some, but not all, of those erroneous situations?

Reply:
> If only one cleaner thread runs, can we just safely assume it is run only once and simply remove the `deallocated` check, which only defends against some, but not all, of those erroneous situations?

Cleaner specifies that it is run once so yes, I think flag can be dropped, as well as checking if the address is 0.

Reply:
`address` can't be 0 if asserts a enabled. Might `address` be 0 if asserts **_aren't_** enabled?

Reply:
This constructor is only used from DirectXBuffer constructors (line 142), and the passed address is the return value of `Unsafe.allocateMemory` (line 129). We can trust `address` as long as we trust the unsafe method.

Reply:
> `address` can't be 0 if asserts a enabled. Might `address` be 0 if asserts **_aren't_** enabled?

No, the address is always non-0. The assert is good, the check if it's 0 can be removed. If you allocateDirect(0) then it will adjusted to allocate 1 byte so the address is non-0.